### DISCUSSION TITLE
Converting a stack for Levure

### DISCUSSION THREAD

**Discussion by seaniepie** [trevordevore/levure#173] [Q&A]
Hi Trevor,
I have a massive stack with about 101 substacks, 805 Cards, 83k objects and 360k lines of code. Your wiki is helpful in letting us know how to start a new project. But what would you recommend as the best way to prepare and convert our application to make use of Levure? 
The original coders of this stack I have inherited had some pretty bad habits of placing a lot of their code into buttons and groups and so a lot of my 'modernising' time is taken moving the code into card and stack scripts and re-referencing all of the links to objects and so on.
Any help you can provide would be appreciated.
Sean

**Comment by macMikey** [comment]
If you want to break your project into behaviors and script-only-stacks, have a look at [Navigator](https://github.com/gcanyon.navigator). It has built-in functionality for doing exactly that. It also let you do test runs with a log before you try for real.
Remember, though, that Levure is a framework, a means for adding features, not for reorganizing your project. You do not need to break your project apart into behaviors/SOS's in order to use Levure. In fact, Trevor has, on different threads, recommended doing exactly that until you have your project working with Levure attached to it before breaking it apart.

**Reply by trevordevore** [reply]
@seaniepie What @macMikey says is what I would recommend. Levure does help you organize your project though. For example, you can stick all of your existing stacks that are displayed to the user in the `ui` folder. If you have library stacks then put those in the `library` folder, etc. That way Levure will automatically load the libraries and make the `ui` stacks available so that you can reference them by stack name without loading them into memory at launch.

Hopefully you could move each of those substacks into it's own stack file in a folder in the `ui` folder that is named after the substack. Then, later on, you can go through and start moving scripts in each stack into behaviors within that folder. This can be more memory efficient when loading the app and if you use the destroyStack property to remove stacks from memory when they close it can keep the memory footprint lower. 

But you don't even have to do that to get started with Levure. You could stick that massive stack in a folder in the `ui` folder and just start there. Levure doesn't care.

**Reply by seaniepie** [reply]
Your link to Navigator doesn't work.
I don't think I am understanding Levure enough. I get that it is more for version control and collaboration. It's a means for getting the codebase onto Github. I wasn't intending to use it entirely for reorganising. But, if I move the project over to Levure, which will benefit me in the long term, I will still need to do a whole heap of reorganising anyway. Having all of these button scripts (which are often repeated over and over because of bad coding prior to me getting hold of it) converted to behaviours would eventually have to be moved over to the SOSs. 
If I make changes to the UI in the IDE, that's in binary and so is any code that is associated with buttons, etc. So how will I reallocate the behavioural reference to the SOS code? It's still a little confusing to me at the moment as I am unfamiliar with the framework.
Where also are the 'different threads' you refer to found?

**Reply by seaniepie** [reply]
I found the link to Navigator. https://github.com/gcanyon/navigator/
But it seems to be abandon-ware. Lot's of issues posted but no pull requests since 2018. It's not the sort of thing I would use if it's not kept up to date and repaired regularly. I couldn't trust it. It's just adding another broken thing on top of LC in a vain effort to replace a similar broken thing (the project browser).

**Reply by macMikey** [reply]
I've used it recently to break a tmc2 project apart into behaviors, so I can confirm that at least for me, that portion works.
Sorry about the broken link.
Also, Monte wrote a component that's buried in the LC bundle that does something similar, called "scriptify", I think.

**Reply by seaniepie** [reply]
Monte's thing is called Scriptifier.livecodescript and is found in the LiveCode/LiveCode/tools repo.
Did you know Navigator 2.5b is _still_ included in the LC9 stack! Bizarre!

**Reply by macMikey** [reply]
Yeah, my Mac has been down doing the latest software update for an hour so I just got in to verify the name, and you already answered.
As you probably have noticed, Navigator 2.5b is not the only old/obsolete thing that comes bundled with LC...

As for editing a stack in the IDE and what happens with the code:
You can assign (and nest) behaviors and have all the code (or as much of it as you like) be separate from your stack file.
Trevor has a number of videos on his [youtube channel](https://www.youtube.com/channel/UCluXVDvheCjGSJmCMssc0fw/videos) - maybe I got the link right, this time. [One of them](https://www.youtube.com/watch?v=eyggLzIbeSU) at the 5:30 mark shows you how to do it manually. You just hit the "behaviors" button in the property inspector (two gears, usually at the far right), and create a behavior from the existing script. Now the script of the object/card/etc. is in a separate .livecodescript file. Now your code and your layout are separate.
